### ISSUE TITLE
[Feature] Ability to pack multiple images into one tar export

### ISSUE THREAD

**Issue by CafeLungo** [regclient/regclient#710]
The `docker image save -o image.tar image:1 image:2 ... ` command supports multiple images to be saved into a single tarball. It would be tremendously helpful if regctl offered this capability because I've been looking for a way to save archives like this (with the deduped layers) without running a docker daemon, and the regclient toolset is very close to enabling this need.

### Current Behavior

regctl image export <image_ref> [filename] [flags]

- Only a single image_ref supported

### Expected Behavior

regctl image export <image_ref...> [filename] [flags]

- Allow multiple image_refs to be defined

### Example Solution

regctl image export <image_ref...> [filename] [flags]

- Allow multiple image_refs to be defined

### Version

```text
(paste output of `regctl version` or similar for binaries, or `docker image inspect regclient/...` on the image and paste the labels)
```

```
VCSTag:     v0.6.0
VCSRef:     9de7397da9f1c00dad5213519366002376b8d5ed
VCSCommit:  9de7397da9f1c00dad5213519366002376b8d5ed
VCSState:   clean
VCSDate:    2024-03-24T18:09:00Z
Platform:   linux/amd64
GoVer:      go1.22.1
GoCompiler: gc
```

### Environment

- Running as binary or container: container
- Host platform: linux/amd64
- Registry description: artifactory in our case, but shouldn't matter

### Anything else

Ref: https://docs.docker.com/reference/cli/docker/image/save/

**Comment by sudo-bmitch** [comment]
Unfortunately, since the filename is optional in that command, I'm not sure of a good way to do this that isn't a breaking change. It can also result in an OCI Layout being output that has some undefined behavior with multiple manifests referenced by the same "tag". This was looked at in #616, without a good resolution.

The best advice I have for this is to copy images to an OCI Layout, each with a unique tag. If copying content across an air-gap scenario, a shared storage directory or rsync is even better because layers that were previously copied would not need to be copied again. An example of what this might look like is:

```shell
# copy into an OCI Layout, flags like "--platform local" can be used to limit how much is copied
regctl image copy ghcr.io/regclient/regctl ocidir://export:regctl
regctl image copy ghcr.io/regclient/regsync ocidir://export:regsync
# copy, with optional tar, of the export directory
regctl image export ocidir://export:regctl --platform local --name ghcr.io/regclient/regctl | docker load
regctl image export ocidir://export:regsync --platform local --name ghcr.io/regclient/regsync | docker load
# or to import into a registry, the copy can be reversed
regctl image copy ocidir://export:regctl ghcr.io/regclient/regctl
regctl image copy ocidir://export:regsync ghcr.io/regclient/regsync
```

**Comment by kohtala** [comment]
The image copy should have option to add the "io.containerd.image.name" annotation such that different images could be copied to the same ocidir.

The offered workaround does not solve the problem of deduplicating the layers. Being able to copy images to same ocidir (or export multiple image in same command) would allow for smaller single tar file to copy to the air-gapped device.

**Comment by sudo-bmitch** [comment]
> The offered workaround does not solve the problem of deduplicating the layers. Being able to copy images to same ocidir (or export multiple image in same command) would allow for smaller single tar file to copy to the air-gapped device.

The contents of the `export` folder in the example above do have the layers deduplicated. Compared to copying a tar file, the directory offers the advantage of being efficient with an `rsync` command that only copies the updated blobs, vs the entire tar file. However it's still possible to tar and compress the directory (note the blobs would typically already be compressed) using the `tar` command.

**Comment by kohtala** [comment]
> The contents of the `export` folder in the example above do have the layers deduplicated. Compared to copying a tar file, the directory offers the advantage of being efficient with an `rsync` command that only copies the updated blobs, vs the entire tar file. However it's still possible to tar and compress the directory (note the blobs would typically already be compressed) using the `tar` command.

Right. My bad reading.

**Comment by kohtala** [comment]
I created PR #857 for the original issue.

I also looked at ocidir to support URI [fragment](https://datatracker.ietf.org/doc/html/rfc3986#section-3.5) to specify the image in the ocidir index. It is partially implemented in #858.

**Comment by sudo-bmitch** [comment]
I'm not sure where to put this since the problem has been spread across three different PRs. Ideally, this should be designed in the issue first before writing the code.

Different tools treat an OCI Layout differently. Some assume they are a single image, possibly only a single platform image, that was export from a builder or docker engine. This is seen with some SBOM and vulnerability scanners that break when they see multiple entries in the index.json file, and it's how builders like buildx output to an OCI Layout.

Other tools treat the OCI Layout as a transport, not usable by itself, but a method of packaging content from a registry or another source. The content is not referenced by the layout itself, but by an identity known outside of the layout. This is how tools like the Docker engine treat it, because even the content inside of the Docker engine is only known by its external name, even if that name is fictitious.

The structure suggested (but not mandated) by the OCI specs is to treat the OCI Layout as a repository. There is a blob store, and the index.json is used to map tags to manifests. This structure is used by regclient which allows images to be accessed in a local OCI Layout identically to how they are accessed on a registry, and the name on the registry isn't important. This makes logical sense to me since the same content can be located in zero to many registry locations, all independent from the layout, and no one name is required to be canonical.

The result of treating an OCI Layout as a repository is that commands like the tag listing, image modification, image copy, image inspection, index manipulation, all work identically whether the content is local or remote. Trying to undo this structural assumption in one place breaks the other uses in unexpected ways that will result in lots of issues demanding fixes in the future. Currently, while there are incompatibilities between different projects that use OCI Layouts, regclient attempts to be internally consistent so that the different uses are all compatible with each other.

This problem would need to be resolved with a structural design decision first, before starting work on the export that would depend on that decision. I understand this is a highly requested use case. But currently fixing this inconvenience for one group would result in breakages and technical debt for many other use cases that I don't wish to incur.

**Comment by kohtala** [comment]
Thank you for taking the time to comment and discuss.

I can appreciate your need to keep the features of the tool consistent. It is your role here and it helps create a better tool. I approach this from a different direction. I have work or a specific task that needs to be done. For me it is not as much about internal consistency, but being able to accomplish the task. The task depends on working together with the other tools. A tar transport is necessary for the work.

As for the many PR's, I am not confident if we first discuss something without coding, I can then deliver an implementation. I'd hate to have a discussion and then disappear when taken onto some other work. I have the task, I can make a change to accomplish that, and after that I prefer spending some time discussing what it would take to be accepted upstream. I'm also motivated by the morals: since I get to enjoy the work done into regclient, I hope to be able to contribute something back and can do a little more for it than required for the task.

I'd have preferred to continue on #857 rather than create the new PR #860, but after #857 was closed, GitHub no longer accepted addressing the concerns you raised in the review comments with a new commit.

My use of regclient would be to create this kind of tar file for transport to be consumed in `docker load`. The regctl image save would be useful, but I expect the workflow could use a staging area before creating the tar file. In this staging area it would be possible to perform some small customization without a new image built. Or merge other tar files into it. ocidir seems like a good storage for that staging. There is no need to have more than one tag for each image/repository, but the images need to load back to docker with the expected repository name.

Since images from different repositories are built on the same base images and share layers and GitHub runners have limited disk space, it is useful to store all the images in the same ocidir. Larger runners cost more. Sometimes also the network bandwidth has a cost and downloading or transferring many tar files with same layers repeatedly has a cost. Bandwidth can be slow, so increased size also costs time. Therefore I need the different images in the same tar.

If using ocidir for staging, feature consistency can be achived with the URI fragments with commands like

- `regctl repo ls ocidir://directory` to list the repositories present in index.json
- `regctl tag ls "ocidir://directory#repository"` to list the tags for the repository present in the ocidir URI as with the #858 URI fragment idea
- `regctl artifact/image "ocidir://directory#repository:tag@digest"` commands with the full URI

If ocifile scheme became implemented, it could achieve feature consistency in the same way. But currently I do not see such feature consistency between ocidir and the tar file.

The `regctl image export` currently sets the ocidir URI into `io.containerd.image.name` annotation, but it could be changed to use just the repository:tag part of the ocidir URI.

I think there is no need for more than one staging directories, so rather than use the ocidir URI fragment, it may be more handy to have a global flag to set the staging directory and use a separate scheme to access it. It could be based on the ocidir implementation and use the index.json. So instead of `ocidir://directory#repository:tag@digest`, it could be an option `--staging directory` and URI like `staging:repository:tag@digest`. Internally the ref could be the same for both schemes, so it could be just a syntactic change on the command line.

In any case, the end result I need would be the tar file to transport the images.

**Comment by sudo-bmitch** [comment]
> If using ocidir for staging, feature consistency can be achived with the URI fragments with commands like
> 
>     * `regctl repo ls ocidir://directory` to list the repositories present in index.json
> 
>     * `regctl tag ls "ocidir://directory#repository"` to list the tags for the repository present in the ocidir URI as with the [Copy many images to ocidir #858](https://github.com/regclient/regclient/pull/858) URI fragment idea
> 
>     * `regctl artifact/image "ocidir://directory#repository:tag@digest"` commands with the full URI
> 
> 
> If ocifile scheme became implemented, it could achieve feature consistency in the same way. But currently I do not see such feature consistency between ocidir and the tar file.

Changing an OCI Layout from a representation of a repository to one of a registry (or multiple registries) would go counter to use cases like Zot and olareg that use the OCI Layout as the backend storage for the registry. As an aside, something I've been looking at for olareg is whether the blob store could be shared between multiple OCI Layout directories. That introduces some complications on things like garbage collection logic, and introduces compatibility challenges between tools, but could have an advantage for disk space usage. Hard links could be used to solve that issue in a way that retains compatibility between tools but limits the platforms where olareg could use that feature.

Is the `ocidir://directory#repository:tag` syntax newly invented for regclient, or a standard already used elsewhere? Adding new characters that can never appear in the directory path has implications for other users that may have a use case for other characters in the path (see #855 and #763) .

There's also potential overlap with proxy namespaces that is being worked in in OCI, and already supported in some registries and containerd. Consistency in a reference format that applies to both would be useful for end users. See https://github.com/opencontainers/distribution-spec/pull/66#issuecomment-2488970063.

Support for `ocifile` is still a long range nice to have in the project. Adding it would replace all of the export/import code, which would be a big net positive (since it is fairly convoluted and difficult to maintain). The downside of adding `ocifile` support is that tar is not a random access format, so all actions would have a significant performance hit, and potentially require lots of disk or memory to avoid repeated seeks through the file. However, if something like estargz or soci were to be developed without CGO, that would change things.
